ipv8/messaging/interfaces/endpoint.py

```diff
@@ -128,11 +130,15 @@ def __init__(self, endpoint, main_thread=True):
 
         self.endpoint = endpoint
 
-        self._netifaces_failed = netifaces is None
+        self._netifaces_failed = self.netifaces is None
         self._local_interfaces = []
         self._my_estimated_lan = None
         self.my_estimated_wan = self.my_estimated_lan
 
+    @property
+    def netifaces(self):  # For easier mocking in tests
+        return netifaces
+
```

Comment:
I'd consider creating mocking code for an entire external dependency the very last resort. It is unnecessary here. Also, since the `EndpointListener` is an ABC you have to make a subclass anyway. Please refactor the `netifaces.interfaces()` call into its own method and use that in `_get_interface_addresses()`.

 For example, in `EndpointListener`:

```python
@classmethod
def get_interfaces(cls):
    return netifaces.interfaces()

@classmethod
def get_interface_addresses(cls, interface):
    return netifaces.ifaddresses(interface)

@classmethod
def _get_interface_addresses(cls):
    ...
    for interface in cls.get_interfaces():
        try:
            addresses = cls.get_interface_addresses(interface)
    ...
```

In the test class you can implement the ABC using an artificial return value for `get_interfaces()`:

```python
class ClassUnderTest(EndpointListener):
    @classmethod
    def get_interfaces(cls):
        return FAKE_ADDRESSES.keys()

    @classmethod
    def get_interface_addresses(cls, interface):
        return FAKE_ADDRESSES[interface]
```

Reply:
Instead of adding these helper methods, I used the `patch` decorator of the `unittest.mock` library, this way I do not overcomplicate the class just for the testing.

Reply:
I think that is an acceptable solution for the target of this PR. However, I consider having to use `patch` to be a code smell of bad architecture (this code is 8 years old after all) and I will open an issue after this PR has been merged.

Reply:
I agree that there are many ways to approach the testability of `_get_interface_addresses`; each has its trade-offs. In my opinion, the solution with class methods is more problematic for the following reasons:
* It continues the general trend used in IPv8 to rely on class inheritance. It is commonly considered better practice to prefer [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) to make code decoupled.
* Class methods are often not flexible enough. If someone needs to write several test cases with different sets of fake addresses, it may be harder to do it with class methods.
* New class methods became part of `EndpointListener` API, complicating it. The naming can become an issue. Do you really want two methods, `get_interface_addresses` and `_get_interface_addresses`, with very similar names but different return types and parameters?
* Why these new methods are public (that is, named without a leading underscore) while they are only called from a protected method `_get_interface_addresses`? Should they be protected methods as well?

Solution with `patch` avoids these issues and looks a bit clearer for me. But it is a matter of personal preferences.

I agree that it is better to refactor the `_get_interface_addresses` method to make it more testable. For example, move the `Interface` class to the top-level to make it testable as well. But it is indeed better to do it in a separate PR.

Reply:
Just because composition is preferred, does not mean inheritance is bad. One is not a better practice than the other. A good architecture applies both to meet its goals. There is more than one design pattern/consideration that must be applied. "Everything must be composable" is a cargo-cult-programming mindset (see also [the list of architecture anti-patterns](https://en.wikibooks.org/wiki/Introduction_to_Software_Engineering/Architecture/Anti-Patterns)).

Class methods are more than flexible enough to do anything short from hyper-composition microservices, which is not how IPv8 is used (and I refer you to [the last Q&A point of IPv8's contributing guidelines](https://github.com/Tribler/py-ipv8/blob/master/CONTRIBUTING.md#common-qa) if you think it is).

> Solution with `patch` avoids these issues and looks a bit clearer for me. But it is a matter of personal preferences.

I see how the resulting unit test code is smaller for the current implementation. However, it does not solve the underlying issue of the code not being easy to test. I don't agree that bad design of the `Endpoint` legacy code is a matter of personal preference. To a certain extent I agree that we should not touch legacy code: "if it works, don't fix it". However, I just linked to a list of anti-patterns that include the "lava flow" anti-pattern: if we start adding more tests, the lack of testability will only make tests worse and worse (even `patch` won't help you then). Then again, will we actually add more tests any time soon? Probably not.

Reflecting your comments---and this thread as a whole---I'll opt not to make an issue. If anyone wants to chase 100% test coverage in the (far?) future, they can solve the testability issue at that point in time. Let's leave the legacy code rest in peace.